cli/src/pcluster/templates/budget_builder.py

```diff
@@ -53,35 +64,141 @@ def _add_parameters(self, budget, index):
+                        threshold_type=single_notification.notification.threshold_type,
+                        threshold=single_notification.notification.threshold,
+                    ),
+                    subscribers=subscribers,
+                )
+            )
+        return notifications_with_subscribers
+
+    def _create_lambda_and_sns(self):
+        """Create an sns topic and a lambda function to stop compute fleet when triggered."""
+        self.sns_topic = sns.Topic(
+            self.scope,
+            "BudgetTopic",
+            display_name="BudgetTriggeredNotification",
+            fifo=False,
+            topic_name=f"{self.cluster_config.cluster_name}-{self.cluster_config.region}",
+        )
+        self.sns_topic.add_to_resource_policy(
+            iam.PolicyStatement(
+                actions=["SNS:Publish"],
+                effect=iam.Effect.ALLOW,
+                resources=["*"],
+                principals=[iam.ServicePrincipal("budgets.amazonaws.com")],
+            )
+        )
+        self.sns_arn = self.sns_topic.topic_arn
+
+        dirname = os.path.dirname(__file__)
+        filename = os.path.join(dirname, "../resources/budget_lambda/lambda_handler.py")
+        with open(filename, encoding="utf8") as lambda_file:
+            handler_code = lambda_file.read()
+
+        lambda_env = {
+            "TABLE_NAME": self.scope.dynamodb_table_status.ref if not self.is_batch else "NONE",
+            "IS_BATCH": "TRUE" if self.is_batch else "FALSE",
+            "CE_NAME": self.scope.scheduler_resources.compute_env.ref if self.is_batch else "NONE",
+        }
+
+        function_id = "BudgetAlert"
+        function_name = f"{PCLUSTER_LAMBDA_PREFIX}{function_id}-{self._stack_unique_id()}"
+
+        budget_triggered_lambda_role = None
+
+        if self.create_lambda_role:
+            slurm_policy = iam.PolicyStatement(
+                actions=["dynamodb:GetItem", "dynamodb:UpdateItem"],
+                effect=iam.Effect.ALLOW,
+                resources=[
+                    self.scope.format_arn(
+                        service="dynamodb",
+                        resource="table/parallelcluster-*",
+                        region=self.scope.region,
+                        account=self.scope.account,
+                    ),
                 ],
+                sid="DynamoDbStopFleetPolicy",
+            )
+            batch_policy = iam.PolicyStatement(
+                actions=["batch:UpdateComputeEnvironment"],
+                effect=iam.Effect.ALLOW,
+                resources=["*"],
+                sid="BatchStopComputeEnv",
+            )
+
+            budget_triggered_lambda_role = add_lambda_role(
+                scope=self.scope,
+                function_id="BudgetTriggeredAction",
+                statements=[batch_policy if self.is_batch else slurm_policy],
             )
+
+        budget_triggered_lambda = lambda_.Function(
```

Comment:
Can we reuse `PclusterLambdaConstruct` to build the lambda function here?

Reply:
Using L1 constructs became a bit impractical when subscribing the lambda to the sns topic.